### PR TITLE
kafka share sarama client between producer and consumers

### PR DESCRIPTION
**Pull request by famarting**
# Description

Changes the way the kafka component initializes the producer and consumer sarama clients, now both will reuse the same underlying sarama.Client , this should reduce the number of requests made on startup to the kafka brokers.

This PR also changes the pubsub functions `GetKafka() *kafka.Kafka` and `GetClient() sarama.Client` so if using this code as a library its possible to reuse the same kafka client for other purposes

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #3336

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by artursouza** [comment]
Is there any disadvantage of sharing the client between pub and sub? If so, it might be good to have it as a config in component metadata to decide to share or not the client.

**Comment by famarting** [comment]
> Is there any disadvantage of sharing the client between pub and sub? If so, it might be good to have it as a config in component metadata to decide to share or not the client.

main disadvantage from my point of view is that each client fetches metadata from all the brokers and for all topics in the cluster, so if you have two clients this process is done twice

**Review by ItalyPaleAle — APPROVED** [review]

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
A bunch of Kafka tests are now failing, can you please take a look?

https://github.com/dapr/components-contrib/actions/runs/7802040724/job/21280858011?pr=3337
https://github.com/dapr/components-contrib/actions/runs/7802040724/job/21280858321?pr=3337
https://github.com/dapr/components-contrib/actions/runs/7802040743/job/21280855464?pr=3337

File: pubsub/kafka/kafka.go [hunk 1]
```diff
@@ -122,6 +122,10 @@ func NewKafka(logger logger.Logger) pubsub.PubSub {
 	}
 }
 
+func (p *PubSub) GetKafka() *kafka.Kafka {
```
Comment (berndverst) [inline pubsub/kafka/kafka.go:125]: Only methods related to the PubSub interface should act on this struct here.

Comment (berndverst) [inline pubsub/kafka/kafka.go:125]: Remove this method.

**Review by berndverst — CHANGES_REQUESTED** [review]
Please remove the method acting on the PubSub struct. Everything else is fine.

I would have rejected the PR ordinarily because publisher and subscriber could should have independent lifecycles, however, I did some digging into the sarama SDK and saw that the ProducerClient under the hood instantiates a regular (producer) client anyway. So given that, this change is totally fine by me.

**Review by berndverst — CHANGES_REQUESTED** [review]
Kafka certification tests must pass before this PR can be merged.

**Comment by famarting** [comment]
> Please remove the method acting on the PubSub struct. Everything else is fine.
> 
> I would have rejected the PR ordinarily because publisher and subscriber could should have independent lifecycles, however, I did some digging into the sarama SDK and saw that the ProducerClient under the hood instantiates a regular (producer) client anyway. So given that, this change is totally fine by me.

could you point me to the place in the code where it instantiates a regular client anyway? ty

File: pubsub/kafka/kafka.go (hunk 1 above)
Comment (famarting) [reply pubsub/kafka/kafka.go:125]: I wanted a way of exposing the internal common implementation of the kafka component for users of this repo as a library, given I added `func (k *Kafka) GetClient() sarama.Client` I was looking for a way of accessing it

Comment (famarting) [reply pubsub/kafka/kafka.go:125]: is there any other approach for achieving this that would be accepted? i.e the entire pubsub kafka implementation could be moved to `common/component/kafka/kafka.go` equivalent to [postgres](https://github.com/dapr/components-contrib/blob/main/state/postgresql/v1/postgresql.go#L24)

which then allows for 
```
st := NewPostgreSQLStateStore(...)
pgst := st.(*postgresql.PostgreSQLQuery)
pgst.GetDB()
...
```

Comment (ItalyPaleAle) [reply pubsub/kafka/kafka.go]: IMHO I think adding a "private" method is fine. We do that in other components and it can be useful for certification tests

**Comment by berndverst** [comment]
> > Please remove the method acting on the PubSub struct. Everything else is fine.
> > I would have rejected the PR ordinarily because publisher and subscriber could should have independent lifecycles, however, I did some digging into the sarama SDK and saw that the ProducerClient under the hood instantiates a regular (producer) client anyway. So given that, this change is totally fine by me.
> 
> could you point me to the place in the code where it instantiates a regular client anyway? ty

Look in the sarama client code within their github repo.

**Pull request closed without merging**